Create `CONTRIBUTING.md`

Philippe-Cholet opened the pull request:
Fixes #756 

As mentioned in the issue, I did not write anything related to testing yet, as I think we should clarify what's expected of contributors.

[Rendered here](https://github.com/Philippe-Cholet/itertools/blob/contributing/CONTRIBUTING.md)

jswrenn commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,91 @@
+
+No guideline about using `#[inline]` yet.
+
+## `core::iter::Iterator` vs `itertools::Itertools`
+For new features, please first consider filing a PR to [rust-lang/rust](https://github.com/rust-lang/rust),
+adding your new feature to the `Iterator` trait of the standard library, if you believe it is reasonable.
+If it isn't accepted there, proposing it for inclusion in `itertools` is a good idea.
+The reason for doing is this is so that we avoid future breakage as with `.flatten()`.
+However, if your feature involves heap allocation, such as storing elements in a `Vec<T>`,
+then it can't be accepted into `libcore`, and you should propose it for `itertools` directly instead.
```
Let's cut this guidance out (and out from our README too). `Iterator` moves slowly enough that conflicts are infrequent, and the acceptance of RFC2845 provides a permanent solution. Given this, I don't feel we need to continue stalling improvements to itertools with this requirement.

Philippe-Cholet replied:
Both removed.

jswrenn commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,91 @@
+The reason for doing is this is so that we avoid future breakage as with `.flatten()`.
+However, if your feature involves heap allocation, such as storing elements in a `Vec<T>`,
+then it can't be accepted into `libcore`, and you should propose it for `itertools` directly instead.
+
```
Let's start with: **File an issue proposing your feature.**

With some text to the effect of: *If you are proposing a major change to CI or a new iterator adaptor for this crate, please first file an issue describing your proposal.*

Philippe-Cholet replied:
I left a comment nearly at the beginning for it to be seen quickly.

jswrenn commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,91 @@
+- Immutably borrow when there is no change, such as `size_hint`.
+
+## Specialize `Iterator` methods
+It might be more performant to specialize some methods.
+However, each specialization should be thoroughly tested.
```
I think we should add something like: *Correctly specializing methods can be difficult, and we do not require that you do it on your initial PR.*

What do you think about this policy, @phimuemue?

phimuemue replied:
@jswrenn I like your idea. (It implicitly aids review by keeping PRs small.)

Should the guidelines mention that we have a canonical way to test specialized methods (in order to avoid a hundred hand-rolled implementations)?

Philippe-Cholet replied:
jswrenn's addition being approved is now committed.
Nothing yet about tests/benchmarks.

Philippe-Cholet commented:
I finally mentionned tests and benchmarks.

jswrenn commented:
Mark this 'ready for review' whenever you feel it's ready to be merged!

Philippe-Cholet commented:
I squashed my commits, I think it's a decent initial `CONTRIBUTING.md` that we will ocasionally update.

CI is blocked since #805. I'm not sure what to do apart from removing the `paths-ignore:` I added in there.

jswrenn commented:
Hm, never mind. That won't help. Honestly, it might just be easiest for me to bypass branch protections for these cases.
